core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1925,14 +1917,34 @@ public void onFailure(Exception e) {
 
                 @Override
                 protected void doRun() throws Exception {
-                    MergePolicy.MergeException e = new MergePolicy.MergeException(exc, dir);
-                    failEngine("merge failed", e);
+                    maybeDie("fatal error while merging", exc);
```

Comment:
Out of curiosity, why `maybeDie` in the generic thread instead of calling `maybeDie` as the first thing in `handleMergeException`?

Reply:
Good question, I should have left a comment explaining this. The reason is because that will be on the Lucene merge thread where I have no guarantees that the call stack does not contain catch throwable! By moving this to another thread that we have complete control over, we know what the stack contains and know this will go uncaught.

Reply:
@dakrone I pushed fb670831c07c4828807b6536442e3f87adcf51bc.

Reply:
See also the Javadoc I had put on `maybeDie` that explains that callers must ensure the call stack does not contain catch statements that would lead to the thrown error being caught and never reaching the uncaught exception handler.

Reply:
Ahh okay, that makes sense. Thanks for the explanation!

What happens if there isn't enough memory to create the new thread? Would it be good to do it in both places just to make best-effort to be sure?

Reply:
It's unlikely that a new thread will be created here, instead that we will simply be reusing an existing one from the generic thread pool. However, it still raises the question of what happens if some allocation in the path fails leading to another out of memory exception being thrown. In that case, that one would go uncaught and still lead to the node being torn down, exactly what we are already trying to achieve.